src/commands/common/get-latest-service-agreement/blockchain-get-latest-service-agreement.js

```diff
@@ -0,0 +1,186 @@
+                );
+                return;
+            }
+            throw error;
+        }
+
+        const latestDbTokenId =
+            (await this.repositoryModuleManager.getLatestServiceAgreementTokenId(blockchain)) ?? 0;
+
+        this.logger.debug(
+            `Get latest service agreement: Latest token id on chain: ${latestBlockchainTokenId}, latest token id in database: ${latestDbTokenId} on blockchain: ${blockchain}`,
+        );
+
+        const missingTokenIds = Array.from(
+            { length: latestBlockchainTokenId - latestDbTokenId },
+            (_, index) => latestDbTokenId + index + 1,
+        );
+        this.logger.debug(
+            `Get latest service agreement: Found ${missingTokenIds.length} on blockchain: ${blockchain}`,
+        );
+        let batchNumber = 0;
+        while (batchNumber * BATCH_SIZE < missingTokenIds.length) {
+            const promises = [];
+            for (
+                let i = batchNumber * BATCH_SIZE;
+                i < missingTokenIds.length && i < (batchNumber + 1) * BATCH_SIZE;
+                i += 1
+            ) {
+                const tokenIdToBeFetched = missingTokenIds[i];
+                promises.push(
+                    this.getAgreementDataForToken(tokenIdToBeFetched, blockchain, contract),
+                );
+            }
+
+            // eslint-disable-next-line no-await-in-loop
+            const missingAgreements = await Promise.all(promises);
+
+            // eslint-disable-next-line no-await-in-loop
+            await this.repositoryModuleManager.bulkCreateServiceAgreementRecords(
+                missingAgreements.filter((agreement) => agreement != null),
+            );
+            batchNumber += 1;
+        }
+        if (missingTokenIds.length !== 0) {
+            this.logger.debug(
+                `Get latest service agreement: Successfully fetched ${missingTokenIds.length} on blockchain: ${blockchain}`,
+            );
+        }
+    }
+
+    async getAgreementDataForToken(
+        tokenId,
+        blockchain,
+        contract,
+        hashFunctionId = CONTENT_ASSET_HASH_FUNCTION_ID,
+    ) {
+        this.logger.debug(
+            `Get latest service agreement: Getting agreement data for token id: ${tokenId} on blockchain: ${blockchain}`,
+        );
+        const assertionIds = await this.blockchainModuleManager.getAssertionIds(
+            blockchain,
+            contract,
+            tokenId,
+        );
+        const keyword = await this.ualService.calculateLocationKeyword(
+            blockchain,
+            contract,
+            tokenId,
+            assertionIds[0],
+        );
+        const agreementId = await this.serviceAgreementService.generateId(
+            blockchain,
+            contract,
+            tokenId,
+            keyword,
+            hashFunctionId,
+        );
+        const agreementData = await this.blockchainModuleManager.getAgreementData(
```

Comment:
since you're getting agreement data here, wouldn't it make sense to also calculate and store bid for each SA ? In that case there would be no need to get agreement data again in epoch check command